### PR TITLE
Ember refactor cicd script

### LINKED ISSUE #39094: [Platform] Ember removal process
### Reproduction steps

To step away from Ember, there are still a lot of files using it in some way or another. The current lint workflow from #38006 aims to evaluate the usage of ember in current files and provide a way to detect it, by slowly removing exclusions. The symbols found that use ember as of today, May 20th 2025, are inside the following file, `symbols.txt`, from which the following list seems like a good place to start:

1. `examples/fabric-bridge-app/fabric-bridge-common/src/BridgedAdministratorCommissioning.cpp:33`
2. `examples/fabric-sync/bridge/src/BridgedAdministratorCommissioning.cpp:33`
3. `examples/tv-app/linux/main.cpp:47`
4. `src/app/app-platform/ContentAppPlatform.cpp:289`
5. `src/app/util/attribute-storage.h:218`
6. `src/darwin/Framework/CHIP/ServerEndpoint/MTRServerEndpoint.mm:327`
7. `examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp:298`
8. `examples/energy-management-app/energy-management-common/device-energy-management/src/device-energy-management-mode.cpp:96`
9. `src/app/app-platform/ContentAppPlatform.cpp:246`
10. `src/app/util/attribute-storage.cpp:224`
11. `src/app/util/attribute-storage.cpp:281`
12. `src/app/util/attribute-storage.cpp:982`
13. `src/data-model-providers/codegen/CodegenDataModelProvider.cpp:302`
14. `src/data-model-providers/codegen/CodegenDataModelProvider.cpp:396`

This is because all of these places apparently have ember only in some comments, however, we still must check thoroughly if some indirect call to ember is taking place or not and either exclude it (but maybe add information) or tackle it.

[symbols.txt](https://github.com/user-attachments/files/20355490/symbols.txt)

### Platform

core (please add to version below)

### Platform Version(s)

1.0.1

### Type

YAML tested

### (Optional) If manually tested please explain why this is only manually tested

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jennygaz**
#### Summary

Fixes parts of #39094 and improves the script for future usage.

#### Related issues

The following files have more prominent ember usage that would need a larger refactor, so they were left out of the PR:

* `src/app/app-platform/ContentAppPlatform.cpp` - Uses ember constructs and calls conditionally
* `src/app/util/attribute-storage.h` - Declares ember calls and uses them extensively
* `src/darwin/Framework/CHIP/ServerEndpoint/MTRServerEndpoint.mm` - Uses ember calls and constructs extensively
* `examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp` - Requires ember for initialization
* `examples/energy-management-app/energy-management-common/device-energy-management/src/device-energy-management-mode.cpp` - Uses ember extensively
* `src/app/util/attribute-storage.cpp` - Uses ember extensively
* `src/data-model-providers/codegen/CodegenDataModelProvider.cpp` - Uses ember extensively

#### Testing

Tested on my own fork with Builds and Lint passing

### CODE REVIEW

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 1]
```diff
@@ -286,7 +286,6 @@ void ContentAppPlatform::SetupAppPlatform()
 
     // Disable last fixed endpoint, which is used as a placeholder for all of the
     // supported clusters so that ZAP will generated the requisite code.
-    // emberAfEndpointEnableDisable(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1)), false);
```
Comment (andy31415) [inline src/app/app-platform/ContentAppPlatform.cpp:289]: comment may also need updating/deleting? It seems to refer to this line of code.

File: examples/tv-app/linux/main.cpp [hunk 2]
```diff
@@ -44,7 +44,6 @@ void ApplicationInit()
     // supported clusters so that ZAP will generated the requisite code.
     ChipLogDetail(DeviceLayer, "TV Linux App: Warning - Fixed Content App Endpoint Not Disabled");
     // Can't disable this without breaking CI unit tests that act upon account login cluster (only available on ep3)
```
Comment (andy31415) [inline examples/tv-app/linux/main.cpp]: this comment also seems to reference the code that is about to get deleted

File: scripts/find_ember_symbols_prefixes.sh [hunk 3]
```diff
@@ -37,9 +37,34 @@ check_symbols() {
...
+    # Get all matches first
+    raw_ember_matches=$(git grep -I -n '\<ember[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
+    raw_emAf_matches=$(git grep -I -n '\<emAf[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
 
-    emAf_matches=$(git grep -I -n '\<emAf[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
+    # Filter out matches that are in comments
+    ember_matches=""
+    if [[ -n "$raw_ember_matches" ]]; then
+        while IFS= read -r line; do
+            # Skip if match is inside a // or /* comment
```
Comment (andy31415) [inline scripts/find_ember_symbols_prefixes.sh]: How are we matching `/*` comments given that they be multi-line?
Generally this code starts being less and less updatable - I have trouble reviewing it in an efficient manner.

Could we instead opt for some linter-style exclusion? Most linters support a way to say "skip this, this is fine" so we could add a keyword like "ember-linter-reference-ember-ok: ...." and let people type their own reasoning. That way we can add relevant descriptions including `TODO: this should be removed` comments in there.

File: scripts/find_ember_symbols_prefixes.sh [hunk 4]
```diff
@@ -37,9 +37,34 @@ check_symbols() {
         esac
     done
 
-    ember_matches=$(git grep -I -n '\<ember[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
+    # Get all matches first
+    raw_ember_matches=$(git grep -I -n '\<ember[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
+    raw_emAf_matches=$(git grep -I -n '\<emAf[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
 
-    emAf_matches=$(git grep -I -n '\<emAf[A-Za-z0-9_]*(' -- './*' "${exclusions[@]}")
+    # Filter out matches that are in comments
```
Comment (bzbarsky-apple) [inline scripts/find_ember_symbols_prefixes.sh]: Long-term we want to remove hits from comments too, no?

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 1 above)
Comment (jennygaz) [reply src/app/app-platform/ContentAppPlatform.cpp:289]: This should be fixed in the latest commit

File: examples/tv-app/linux/main.cpp (hunk 2 above)
Comment (jennygaz) [reply examples/tv-app/linux/main.cpp]: This is also fixed in the latest commit

File: scripts/find_ember_symbols_prefixes.sh (hunk 3 above)
Comment (jennygaz) [reply scripts/find_ember_symbols_prefixes.sh]: For now, what I did was simply match in the same line so if we have a line that starts with either `//` or `/*` and it contains `ember`, `emAf`, or similar then it shouldn't trigger it as a positive. To avoid matching symbols inside comments I think a syntactic analysis could be made before or after the other matches and check properly which ones shouldn't be there. Should I add this now in this PR?

File: scripts/find_ember_symbols_prefixes.sh (hunk 4 above)
Comment (jennygaz) [reply scripts/find_ember_symbols_prefixes.sh]: Yes, in the end we want to remove all hits from comments and this PR filters some of them though not all (e.g., when a line with an ember hit is not the same as when a comment starts or ends)

Comment (bzbarsky-apple) [reply scripts/find_ember_symbols_prefixes.sh]: No, I mean don't we eventually want to remove all mentions of "ember" from comments?

Comment (jennygaz) [reply scripts/find_ember_symbols_prefixes.sh]: Yes, though that is going to be by the time ember is removed as a dependency too. In this PR, there are none addressed besides false positives. However, in that case, I could roll back the changes to the script so that comments are still marked for now.

File: scripts/find_ember_symbols_prefixes.sh (hunk 3 above)
Comment (andy31415) [reply scripts/find_ember_symbols_prefixes.sh]: I think once we start having `IFS` it is a hint of `this is too complex for shell`.
My feedback would be to keep this simple as before and allow a "no-lint" comment equivalent. Then instead of an exclude list, we start adding comments and issue links in files.

**Review by andy31415 — APPROVED** [review]
Changes in comment removal look good. I am concerned that the shell script starts being too complex for shell and is incomplete in general (no multiline comment support, but almost supporting them).

File: scripts/find_ember_symbols_prefixes.sh (hunk 3 above)
Comment (andy31415) [reply scripts/find_ember_symbols_prefixes.sh]: I wonder if this handles edge cases of `/* emberAf` should not match, but `/* test */ emberAf` should match ... the greedyness of grep is generally tricky.

Did not actually fully review/test this line because of that. I would prefer simple matches that is "obviously correct".

Comment (andy31415) [reply scripts/find_ember_symbols_prefixes.sh]: Just tested and `/* emberAfTest(` matches (good) but so does `/* */ emberAfTest(` ... so we are obviously not perfect and I believe we should not try to. Trying to detect comments seems painful.
